MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -2023,30 +1969,38 @@ SeparateSpecialRegionsInMemoryMap (
         // If the special region ends before the end of this descriptor region, insert a new record at the end
         // of the memory map for the remaining region
         if (SpecialRegionEnd < MapEntryEnd) {
-          // Populate a new descriptor for the region after the special region
+          // SpecialRegionStart is now guaranteed to be equal to MapEntryStart. Populate a new descriptor
+          // for the region covered by the special region. This entry needs to go to
+          // the end of the memory map in case a subsequent special region will cover some portion
+          // of the remaining map entry region
           POPULATE_MEMORY_DESCRIPTOR_ENTRY (
             MapEntryInsert,
-            SpecialRegionEnd,
-            EFI_SIZE_TO_PAGES (MapEntryEnd - SpecialRegionEnd),
+            SpecialRegionStart,
+            EFI_SIZE_TO_PAGES (SpecialRegionEnd - SpecialRegionStart),

```

Comment:
I'm slightly confused by the comment vs code here. The comment says this goes to the end of the memory map, in the code it goes to the end of the special region. Is that guaranteed to be the end of the memory map? Or am I misinterpreting?

Reply:
It is a bit confusing. In this context, "end of the memory map" means "the next available blank EFI_MEMORY_DESCRIPTOR at the end of the memory map."

Explained another way: If I find that the memory range described by the special region ends before the end of the memory range described by the memory map descriptor I'm currently inspecting, then I want to create a new memory map descriptor that describes the special region and put that descriptor at the end of the map. Then, I trim the memory map entry I'm currently inspecting to just the remaining region uncovered by that special region in case a subsequent special region is also covered by that remaining region.

This is one of those things that's easier with a visual representation, but my ASCII art is subpar :D